### DISCUSSION TITLE
B站最新这期Using+IDisposable的下划线弃元我觉得有点问题

### DISCUSSION THREAD

**Discussion by ali50m** [BYJRK/DotNet-Discussions#294] [C# 技术交流]
@BYJRK 

我觉得在视频[借助Using+IDisposable实现类似Go语言的defer延迟执行效果](https://www.bilibili.com/video/BV1Ym421T7CS?t=353.5)中,不应该使用下划线替换掉变量名md. 

测试代码如下, 我在MyDisposable 中增加了一个公开的测试方法Bar, 会发现`Test1`中下划线能调用Bar.我觉得正常的下划线使用应该类似`Test2`中, 下划线是没办法调用string.Length的.

```csharp

public partial class MainWindow
{
    public MainWindow()
    {
        InitializeComponent();
    }

    private void Test1()
    {
        using var _ = new MyDisposable(() => Console.WriteLine("Job is done."));
        var t = _.Bar(); // 这里可以调用_
        var length = t.Length;
    }

    private void Test2()
    {
        using var md = new MyDisposable(() => Console.WriteLine("Job is done."));
        _ = md.Bar();
        var length = _.Length; // 这里无法调用_,编译失败
    }
}

public class MyDisposable : IDisposable
{
    private readonly Action _callback;

    public MyDisposable(Action callback)
    {
        _callback = callback;
    }

    public void Dispose()
    {
        _callback();
    }

    public string Bar()
    {
        return "hello";
    }
}
```

![image](https://github.com/BYJRK/DotNet-Discussions/assets/9393831/a5078431-c460-4459-b401-2b6356531879)

**Comment by ali50m** [comment]
PS: `var _` 我觉得这是定义了一个叫下划线的变量...

**Reply by BYJRK** [reply]
你说得对，这里确实是声明了一个名为下划线的变量，毕竟这个变量名也是符合规范的。你可以去视频评论区也提一下这个

**Reply by ali50m** [reply]
这样使用下划线变量名的话, 容易和弃元Discards混淆了, 还会造成后续代码块无法正常使用弃元.

**Reply by BYJRK** [reply]
我去视频下方发了条评论，提醒大家一下

**Reply by baciZzz** [reply]
这语法不支持弃元感觉有点可惜啊，有什么不得已的原因吗

**Comment by KnightYyj** [comment]
例子很好，很好的解释了using的用法，但是我觉得 dispose还是偏向释放资源或者加锁解锁这种操作，你这样的骚操作看着是可以，但是不推荐吧。

**Comment by tsctsc6** [comment]
弃元是忽略表达式的结果。由于 using 一个东西，到最后是要调用 Dispose 方法，所以 using 的东西绝不能是弃元。
另外，我建议在使用 defer 功能时，可以把变量起名为 `_defer0_` 这种像是编译器自动生成的样子，防止命名冲突。
